ribs/archives/_unstructured_archive.py

```diff
@@ -0,0 +1,524 @@
+            "objective": <dtype>, "measures": <dtype>}``.
+        extra_fields (dict): Description of extra fields of data that is stored
+            next to elite data like solutions and objectives. The description is
+            a dict mapping from a field name (str) to a tuple of ``(shape,
+            dtype)``. For instance, ``{"foo": ((), np.float32), "bar": ((10,),
+            np.float32)}`` will create a "foo" field that contains scalar values
+            and a "bar" field that contains 10D values. Note that field names
+            must be valid Python identifiers, and names already used in the
+            archive are not allowed.
+    """
+
+    def __init__(self,
+                 *,
+                 solution_dim,
+                 measure_dim,
+                 k_neighbors,
+                 sparsity_threshold,
+                 learning_rate=None,
+                 threshold_min=-np.inf,
+                 qd_score_offset=0.0,
+                 seed=None,
+                 dtype=np.float64,
+                 extra_fields=None):
+
+        ArchiveBase.__init__(
+            self,
+            solution_dim=solution_dim,
+            cells=0,
+            measure_dim=measure_dim,
+            learning_rate=learning_rate,
+            threshold_min=threshold_min,
+            qd_score_offset=qd_score_offset,
+            seed=seed,
+            dtype=dtype,
+            extra_fields=extra_fields,
+        )
+
+        self._k_neighbors = int(k_neighbors)
+        self._sparsity_threshold = np_scalar(sparsity_threshold,
+                                             dtype=self.dtypes["measures"])
+
+    @property
+    def k_neighbors(self) -> int:
+        """int: The number of nearest neighbors to use for determining
+        sparsity."""
+        return self._k_neighbors
+
+    @property
+    def sparsity_threshold(self) -> float:
+        """:attr:`dtype` : The degree of sparseness in metric space required for
+        a solution to be added unconditionally."""
+        return self._sparsity_threshold
+
+    def index_of(self, measures, resize: bool = False) -> np.ndarray:
+        """Returns archive indices for the given batch of measures.
+
+        First, if the archive is empty it is resized to fit the new data.
+
+        Then the distance between all solutions in the archive and the incoming
+        `measures` is calculated.
+
+        Next the distances between all the incoming measures and themselves is
+        also calculated and concatenated with the distances from before.
+
+        Next, we compute the mean distance of the k-nearest neighbors for each
+        incoming measure. If any of the mean distances are *above* the
+        `sparsity_threshold` then the archive is resized to accomodate them.
+        They are each assigned new indices in the archive.
+
+        Args:
+            measures (array-like): (batch_size, :attr:`measure_dim`) array of
+                coordinates in measure space.
```

Comment:
What does resize do? I'd prefer not to change the API of index_of if possible.

Reply:
`resize` indicates whether or not the archive is allowed to grow to support any solutions which match the novelty criterion, or whether or not to simply return the index of the nearest neighbor.

I did this partly so that the archive didn't grow in memory unless we specifically intended to add solutions. I too didn't want to change the API in this way so I can remove the check.